scripts/globals/pets/avatar.lua

```diff
@@ -0,0 +1,501 @@
+    local lastCastTime = pet:getLocalVar(lastCastTimeVar)
+    local lastCastTimeStamp = pet:getLocalVar(lastCastTimeStampVar)
+    if
+        lastCastTimeStamp > 0 and
+        os.time() - lastCastTimeStamp > lastCastTime
+    then
+        lastCastTime = os.time() - lastCastTimeStamp
+    end
+
+    if
+        lastCastTime > 0 and
+        pet:getCurrentAction() ~= xi.action.MAGIC_CASTING
+    then
+        pet:setLocalVar(lastCastTimeStampVar, 0)
+        pet:setLocalVar(lastCastTimeVar, lastCastTime)
+    end
+
+    pet:setMobMod(xi.mobMod.MAGIC_COOL, lastCastTime + castingCooldown)
+end
+
+xi.pets.avatar.onMobSpawn = function(pet)
+    local master = pet:getMaster()
+    if
+        not master or
+        master:getObjType() ~= xi.objType.PC
+    then
+        return
+    end
+
+    -- add listener to player to fine-tune spirit pact cast delays in realtime
+    if
+        pet:getPetID() <= xi.petId.DARK_SPIRIT
+    then
+        -- stops the pet from immediately casting a spell on spawn and respecting the cooldowns by exiting early if MAGIC_COOL is 1
+        pet:setMobMod(xi.mobMod.MAGIC_COOL, 1)
+        pet:setMod(xi.mod.MPP, 500)
+        pet:updateHealth()
+        pet:setMP(pet:getMaxMP())
+
+        master:addListener('TICK', playerListenerVar, function(playerArg)
+            local petArg = playerArg:getPet()
+
+            if petArg and petArg:getMobMod(xi.mobMod.MAGIC_COOL) > 1 then
+                setMagicCastCooldown(petArg)
+            end
+        end)
+
+        master:addListener('ABILITY_USE', playerListenerVar .. 'ABILITY', function(playerArg, target, ability, action)
+            local petArg = playerArg:getPet()
+            local abilityID = ability:getID()
+
+            if
+                petArg and
+                (abilityID == xi.jobAbility.ASSAULT or
+                abilityID == xi.jobAbility.RETREAT)
+            then
+                printDebug(petArg, 'resetting cast cooldown')
+                -- reset cast cooldown via same method as fresh spawn
+                petArg:setMobMod(xi.mobMod.MAGIC_COOL, 1)
+                petArg:setLocalVar(buffModeVar, 1)
+            end
+        end)
+    end
+end
+
+xi.pets.avatar.onMobDeath = function(pet)
+    local master = pet:getMaster()
+
+    if master and master:getObjType() == xi.objType.PC then
+        master:removeListener(playerListenerVar)
+        master:removeListener(playerListenerVar .. 'ABILITY')
+    end
+end
+
+xi.pets.avatar.onMobMagicPrepare = function(pet)
+    -- Note that:
+    -- returning -1 in this function forces TryCastSpell to exit without choosing/casting a spell, but
+    -- will still set the m_LastMagicTime to ensure next call of this function is after the cast delay
+    -- Also, if we return nothing (or zero) TryCastSpell will default to normal mob casting behavior (nukes from spell list, etc)
+    printDebug(pet, 'onMobMagicPrepare: ' .. pet:getMobMod(xi.mobMod.MAGIC_COOL)) -- for debugging magic cooldown
```

Comment:
For all of your debug printing, can you use `string.format` or the new `fmt` helper? I _believe_ I was talking to Teo recently about Lua concat being unsafe, but I don't remember the details. Also Lua concat makes my eyes bleed

Reply:
Fair enough, that was one of those temp-to-perm debug prints

Reply:
Done.

Reply:
yeah concat'ng things can lead to crashes in some cases, that fmt or `string.format()` will catch and turn into neat little "here's what you did vs what I was expecting" messages